src/compiler/moduleSpecifiers.ts

```diff
@@ -844,17 +847,50 @@ const enum MatchingMode {
     Pattern,
 }
 
-function tryGetModuleNameFromExportsOrImports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode: MatchingMode): { moduleFileToTry: string; } | undefined {
+function getOutputPath(targetFilePath: string, options: CompilerOptions, commonDir: string) {
+    if (options.emitDeclarationOnly) {
+        return;

```

Comment:
Where did this check come from? I feel like it might make sense, for auto-imports at least, to assume that an option like this may only be enabled temporarily, or is being used because Babel is emitting JS files, so we should not take it to mean that a JS file can’t exist—maybe we should assume that `outDir` is still going to be accurate in the end.

Reply:
> Where did this check come from? 

I based this on:
https://github.com/microsoft/TypeScript/blob/60ce788302f75a136ce327f8b7f2499ea5dd96f0/src/compiler/emitter.ts#L602-L603

> maybe we should assume that outDir is still going to be accurate in the end.

It's kind of a big bet. OTOH, if somebody is using `package.json#imports` and referencing output files there... then maybe they want this? But dunno, this whole remapping thing wasn't ever intuitive to me and I'm not sure if people know how this works. I only learned how this works and how I could not worry about imports meaning one thing in sources and one thing in dist files when I asked you about this months back. My first instinct was that I'd have to use conditions or smth to "route" those correctly and not that TS would remap this automatically for me.

When I was using Babel I'd typically emit files with it to a different directory then I'd use TS to emit declarations to `types` directory. If I'm not mistaken, at that point I wouldn't even bother about configuring `outDir` (why would I? I wasn't using `tsc` to emit JS files so it didn't have to know the target location). 

Similarly, when I'm using various Rollup-based builders... I just bundle in `package.json#imports`. That has a cost since I need to create a bundle per conditions combination but that's a different topic. With those I can't even easily (I think?) deliver some output that TS would understand here. If i'd put those as `external` in Rollup then it wouldn't even touch them - but we need to touch them since we want to transpile them and everything. Even if I could easily mark them as chunks to be preserved... then this still wouldn't play well with those setups since Rollup heavily relies on the dependency graph. So I wouldn't be able to somehow "include" different variations of the same private module and for a single Rollup build to deal with it.

So overall... I think it's quite a gamble to assume that those files might be in the correct place. But maybe it would be OK in practice solely based on the fact that it might be somewhat convoluted to create a scenario in which this completion is offered when it shouldn't. I suspect that in most of those other cases, it just wouldn't be able to find anything meaningful and thus it wouldn't offer those completions. But at those times - when it would be able to offer a completion maybe it would be the correct one in the vast majority of the cases?

Reply:
I generally think `noEmit` and `emitDeclarationOnly` should affect as little as possible outside of whether certain files actually appear on disk. For example, setting `noEmit` doesn’t disable being able to resolve module specifiers like `"./foo.js"` to `./foo.ts`. The helper in the emitter returns `undefined` because it’s actually used to determine whether to emit something. In this code path, I think we can ignore it.

Reply:
Ok 👍  I just pushed out this requested change